modin/core/io/text/text_file_dispatcher.py

```diff
@@ -1023,24 +1040,46 @@ def _read(cls, filepath_or_buffer, **kwargs):
             )
 
         is_quoting = kwargs["quoting"] != QUOTE_NONE
+        usecols = kwargs["usecols"]
         use_inferred_column_names = cls._uses_inferred_column_names(
-            names, skiprows, kwargs.get("skipfooter", 0), kwargs.get("usecols", None)
+            names, skiprows, kwargs["skipfooter"], usecols
         )
 
-        pd_df_metadata = cls.read_callback(
-            filepath_or_buffer_md,
-            **dict(kwargs, nrows=1, skipfooter=0, index_col=index_col),
+        compute_metadata_before_skipping_rows = (
+            # basic supported case: isinstance(skiprows, int)
+            not isinstance(skiprows, int)
+            # complex cases
+            or (usecols is not None and skiprows is not None)
+            or pre_reading != 0
         )
-        column_names = pd_df_metadata.columns
-        column_widths, num_splits = cls._define_metadata(pd_df_metadata, column_names)
+        read_callback_kw = dict(kwargs, nrows=1, skipfooter=0, index_col=index_col)
+        if compute_metadata_before_skipping_rows:
+            pd_df_metadata = cls.read_callback(
+                filepath_or_buffer_md,
+                **read_callback_kw,
+            )
+            column_names = pd_df_metadata.columns
+            column_widths, num_splits = cls._define_metadata(
+                pd_df_metadata, column_names
+            )
+            read_callback_kw = None
+        else:
+            read_callback_kw = dict(read_callback_kw, skiprows=None)
+            # `memory_map` doesn't work with file-like object so we can't use it here.
+            # We can definitely skip it without violating the reading logic
+            # since this parameter is intended to optimize reading.
+            # For reading a couple of lines, this is not essential.
+            read_callback_kw.pop("memory_map", None)
+            # These parameters are already used when opening file `f`,
+            # they do not need to be used again.
+            read_callback_kw.pop("storage_options", None)
+            read_callback_kw.pop("compression", None)
 
         # kwargs that will be passed to the workers
         partition_kwargs = dict(
             kwargs,
             fname=filepath_or_buffer_md,
-            num_splits=num_splits,
             header_size=0 if use_inferred_column_names else header_size,
-            names=column_names if use_inferred_column_names else names,
```

Comment:
Can we just move the definition of `partition_kwargs` lower to where these parameters are already available? It's better to have this dict one-time defined rather than silently pushing new parameters after its construction.

Reply:
done